app/src/main/java/ch/epfl/sdp/appart/database/local/LocalDatabase.java

```diff
@@ -0,0 +1,1055 @@
+     * or on disk.
+     *
+     * @return the list of cards, or null if it doesn't find it
+     */
+    public List<Card> getCards() {
+        return getFromMemory(() -> cards);
+    }
+
+    /**
+     * Returns an ad with ad id.
+     *
+     * @param adId the id of the ad
+     * @return the ad if it finds it, null otherwise
+     */
+    public Ad getAd(String adId) {
+        return getFromMemory(() -> idsToAd.get(adId));
+    }
+
+    /**
+     * Returns a user with the user id as its id.
+     *
+     * @param wantedUserID the user id
+     * @return the user if it finds it, null otherwise
+     */
+    public User getUser(String wantedUserID) {
+        return getFromMemory(() -> idsToUser.get(wantedUserID));
+    }
+
+    /**
+     * This reads the folder of a user. It will load the user data from disk.
+     *
+     * @param userFile a file pointing to the user's directory
+     * @return a boolean that indicates if the operation succeeded or not
+     */
+    private boolean readUserFolder(File userFile) {
+
+        String dataPath =
+                new StoragePathBuilder().toDirectory(userFile.getPath()).withFile(LocalDatabasePaths.userData);
+        Map<String, Object> userMap = FileIO.readMapObject(dataPath);
+
+
+        User user = UserSerializer.deserialize((String) userMap.get(ID),
+                userMap);
+        if (!idsToUser.containsKey(userMap.get(ID))) {
+            this.idsToUser.put((String) userMap.get(ID), user);
+            this.userIds.add(user.getUserId());
+        }
+
+        return true;
+    }
+
+    /**
+     * This reads the whole user data stored for the current user. It traverses
+     * all the folder with a user id as their name, more specifically all the
+     * folders in the users folder.
+     *
+     * @return a boolean that indicates if the operation succeeded or not
+     */
+    private boolean readUsers() {
+        String userPath =
+                LocalDatabasePaths.usersFolder(getCurrentUser().getUserId());
+        File userFolder = new File(userPath);
+        boolean success = true;
+        for (File folder :
+                Objects.requireNonNull(userFolder.listFiles(File::isDirectory))) {
+            success &= readUserFolder(folder);
+        }
+        if (success) {
+            this.firstLoad = true;
+        }
+        return success;
+
+
+    }
+
+    /**
+     * This function is called every time we want to get something from the
+     * local db. It checks if the data we have in memory is valid, if it is
+     * it directly returns it. Otherwise, it loads everything on disk, and if
+     * the operation succeeded, returns the wanted data.
+     *
+     * @param returnFunc the function that performs the actual retrieving of
+     *                   data from one of the data structures of the class.
+     * @return the data the user wants, or null if it doesn't find it
+     */
+    private <T> T getFromMemory(Supplier<T> returnFunc) {
+        if (this.firstLoad) {
```

Comment:
So, this method is called by all the getters of the class. Since the code for each of them was similar I decided to create this method. returnFunc is typically this.cards.get(cardId) so it gets the values in memory.

The variable is used to prevent unnecessary loading from disk, if we are always online for instance. Once the user has performed one read we are now in "memory mode" and every time an ad is written on disk we will also update the data in memory. Therefore, once firstLoad is true we can always get the value in memory.